### DISCUSSION TITLE
CSLA 5 + Asp.Net Core API

### DISCUSSION THREAD

**Discussion by RocketLabsDev** [MarimerLLC/csla#2742] [Questions]
Hello,

We are starting up with CSLA 5 and have a database context that is created by giving the constructor the connection string.
```
        public Context(string connectionString)
        {
            Initialize(connectionString);
        }
``` 
We are using DI to inject it to the DAL and the context is used in every DataAccess method.
```
namespace DataAccessLayer
  {
      public class VoucherDal : IVoucherDal
      {       private Context Context { get; }
          public VoucherDal(IContext context)
          {
              Context = (Context)context;
          }
```

This now needs to be added to the service collection:
```
  public static class ConfigurationExtensions
  {
    /// <summary>
    /// Add the services for SDK DAL
    /// </summary>
    /// <param name="services"></param>
    public static void AddDalSDK(this IServiceCollection services, string server, string connectionString)
    {
      services.AddScoped<IContext>(ctx => new SDK.Core.Context(connectionString));
      services.AddTransient<IVoucherDal, VoucherDal>();
      services.AddTransient<IMTDal, MTDal>();
    }
  }
```
In the test program it is odone like this and that works:
```
        static void Main(string[] args)
        {
            var services = new ServiceCollection();
            services.AddDalSDK("Server=myServerAddress;Database=myDataBase;Trusted_Connection=True;");
            services.AddCsla();
            VoucherTest();
```
But how is this done in ASP.Net Core API
This connection string should change on every session or request?
Where does the AddCsla come in?
Where should the db context be created?
Program.cs??
```
var builder = WebApplication.CreateBuilder(args);

// Add services to the container.

builder.Services.AddControllers();
// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();


var app = builder.Build();

// Configure the HTTP request pipeline.
if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
}

app.UseHttpsRedirection();

app.UseAuthorization();

app.MapControllers();

app.Run();
```

**Comment by rockfordlhotka** [comment]
I've read a lot of posts and SO threads about how to manage db connections with different connection strings within the context of DI (aspnetcore). The most consistent answer seems to be that people create different types for the different connection "types", and then request the needed type for injection.

**Comment by rockfordlhotka** [comment]
Regarding the call to `AddCsla`, in CSLA 5 the `builder.Services.AddCsla();` call should be the last thing, right before `builder.Build`.

In CSLA 6 we did a better job, and the `AddCsla` call can go anywhere in the `builder.Services` area.

In CSLA 5 there's also an `app.UseCsla` call that's required. This is gone in CSLA 6, but is required in CSLA 5.

**Comment by RocketLabsDev** [comment]
Thank you for the answers!

The problem with the dbContext is that it has to be added to the service collection at startup but how do you modify it when a new request comes in?
It is not that it is different types, it is that there are different databases for every user, and it can't be done so every users dbContext is created when the app is started, then the service would need to be restarted when a new user is added.
The users have a token or hash that is sent in the header of every request and that token/hash would translate to db and user.
An example or idea how this could be done would be much appreciated.

**Reply by rockfordlhotka** [reply]
I honestly don't know. That's a good question for stackoverflow or some forum focused on aspnetcore.

**Reply by RocketLabsDev** [reply]
Ok, thank you and good job on the csla framework.
Will try stackoverflow for the aspnet problem.
